start on archiving behavior spec doc (#402)
piwik/developer-documentation: docs/4.x/archiving-behavior-specification.md

---
category: API Reference
---

# Archiving Behavior Specification

This page describes all the behavior of the internal report aggregation and storing mechanism (archiving)
and the expected behavior of the `core:archive` command (the cron archiving process).
It serves as a reference since the number of cases to think about is too high to keep in ones head, and as
a jumping off point for manual testing of the whole process.

It will be added to over time, since this process in Matomo is especially complex.

## Core Archiving

_Brief Description of system:_ the entry point for this is the `Piwik\Archive` class which handles both querying
for archive data, and launching the report aggregation process if archive data isn't found. `Piwik\ArchiveProcessor\Loader`
is in charge of determining if archiving should be done and then launching the whole process. `PluginsArchiver` is
the entrypoint for report aggregation.

### General Expected Behavior

API methods use the `Piwik\Archive` class to query for archive data. If we are allowed to archive in the current
request/process, and cannot find a recent, usable archive, we generate it, by launching the core archiving process.

### Settings

> * browser triggered archiving: this setting determines whether log aggregation is allowed to be launched from browser requests.

Reviewer: maybe could replace `from browser requests` with `this setting determines whether log aggregation is allowed to be launched on demand when the data is requested by a browser or another client` or something similar?

Reviewer: just mentioning it as I always found browser archiving confusing and not clear. It's more like browser archiving == on demand archiving and cli archiving == regular background archiving or so.

Author: :+1:

>   If disabled, this only happens if a segment that is not set to auto-archive is requested (referred to as a custom segment), or

Reviewer: ```suggestion
  If browser archiving is disabled, archiving only happens if a segment that is not set to auto-archive is requested (referred to as a custom segment), or
```

Author: :+1:

  archive is older than this value, then it is considered outdated. Controlled by a UI setting and the
  `[General] time_before_today_archive_considered_outdated` INI config option. This is also used as the default value
  for archives for other periods.
* `[General] time_before_week_archive_considered_outdated`, `[General time_before_month_archive_considered_outdated]`,
  `[General] time_before_year_archive_considered_outdated`, `[General] time_before_range_archive_considered_outdated`:
  specific ttls for different period types. They each default to the 'today time to live' value if not specified
* custom date ranges to pre-process: there is an INI config setting and some user settings that allow users to specify that
  certain ranges should be pre-archived. The INI setting is `[General] archiving_custom_ranges`. The user setting is the
  setting that controls the default period to load in Matomo. These ranges will be processed in core:archive if specified.

### Rules

**When archiving is allowed/disallowed for the current process**

If browser archiving is enabled, any HTTP request or CLI command can initiate archiving. Things are more complicated

> if browser archiving is disabled:

Reviewer: we might need to mention the configured archiving TTL? Or should it be mentioned in `**Optimizations that skip archiving**`? the ttl applies to browser archiving enabled and disabled. TTL is only looked at I think when date includes today?

also for ranges AFAIK we check the TTL etc.

Author: I don't think the TTL is really about when archiving is allowed/disallowed, it's about when is it necessary/unnecessary. I guess it could be Optimizations. It's described below as well.

Reviewer: I reckon it's important to be aware of for someone not knowing about it (and it also defines whether archiving will be happening or not in a current process). Either here or optimisations is fine. We could just add a sentence with a link to the section that mentions TTL in more details.

Author: :+1:

**Optimizations that skip archiving**

_When there are no visits_

In certain situations, Matomo will be able to tell it does not have to archive and will skip it for a site and period.
This is generally because the site has no visits for that period, and we are able to easily tell.

The specific logic is in `Loader::canSkipThisArchive()`. If a website/period pair is:

- has no visits in the log tables for the period
- and has no subperiod archive within the period (ie, if we're looking at a week and there are no day archives for that week, this would be true)

We know we'd end up with zero visits if we archive, so we can just skip it. The only exceptions are for sites that do
not use the tracker (so there will always be zero visits in the log table) or are specified via the `'Archiving.getIdSitesToArchiveWhenNoVisits'`
event. RollUp websites are examples of sites for which this optimization does not apply.

This optimization is used both before launching the archive aggregation logic and in the `core:archive` command
before we launch individual archive requests. This saves a bit more time since we also don't have to launch an
archiving command.

> ### Special Query Parameter Handling

Reviewer: ```suggestion
### Special URL Query Parameter Handling
```

Author: :+1:

will both: insert into this table before processing each site, and then process the entries, launching the core archiving process
for each valid entry.

### Range Archiving vs All Others

By default when using core:archive, day, week, month & year periods are not processed in the browser. Range periods
on the other hand are still processed from web requests. This is because we cannot pre-archive every possible permutation
of dates that users can specify.

Instead when a range period is no longer considered up-to-date (it's TTL expires), we launch archiving for the range.
Since the range just aggregates day periods, this is still performant, as long as archiving for the days is not launched.

### Launching the core archiving process

The core:archive command launches the core archiving process through new processes, which allows it to achieve concurrency.
If it's supported by the OS configuration and PHP runtime configuration, these processes are CLI processes spawned through
`shell_exec()` running the `climulti:request` command. The `climulti:request` command takes a query string and processes it
like it is an HTTP request.

> If it's not supported, then we use CURL multi requests. CURL is not as fast as launching cli processes, so we prefer to use

Reviewer: I think it's as fast but the problem was that web servers often have request timeouts configured causing the archiving request to be aborted eg after 30 seconds, or after 10 minutes, etc. 

Another bonus was that cli archiving allows us to know how many archivers are currently running (eg across accounts) in the background by executing `ps` etc.

Author: I remember it saving ~15% per request compared to CURL. Not sure why or whether this was on different hardware (it was one back when piwikpro was around I think so can't remember). Will change it though.

`climulti:request`.

In each case, we call the `CoreAdminHome.archiveReports` API method which launches core archiving.
Note: previously the core:archive command would query `API.get` which would implicitly trigger the archiving process.

The request URI will always have `trigger=archivephp` which, when superuser access is granted as is the case during archiving,
will force the archiving process to initiate.

**Batching Archive Jobs**

In the core:archive command, we try and process multiple archives at the same time to make the process more performant. The
number of concurrent processes to run at a time is specified by the `--concurrent-requests-per-website` CLI option.

Archive jobs for different sites are never run concurrently from within the same `core:archive` process. Sites can be
archived concurrently but this is only possible by running multiple `core:archive` commands.

**Order of Execution**

> Some archive invalidations have to be handled before others. Specifically:

Reviewer: could mention here the "why". Eg all visits might be executed first because it is viewed by default and looked at most of the time so this data is visible in the UI sooner. Also more recent dates are viewed more often compared to older dates which is why we archive these first

Author: Was it also to avoid some unpredictability in archiving?

Another core:archive process may then take the next site and process that one in parallel.

If `--force-idsites` is used, core:archive will not use SharedSiteIds but FixedSiteIds. This class has no knowledge of
SharedSiteIds, so it is possible by using this we can launch archiving for the same site multiple times in parallel,
which is generally something we want to avoid.

### General Expected Behavior

**invalidating today & yesterday archives**

`core:archive` will automatically invalidate the archives for `today` and `yesterday` if certain criteria are met.

For `today`, the archive will be automatically invalidated if there are visits for today and if the existing archive is
older than the configured TTL (see [the previous section on archiving settings](#settings)).

For `yesterday`, the archive will be automatically invalidated if the current date is a different day than the `ts_archived`
date for the latest archive for `yesterday`. If this is true, it means the day has changed since the archive was processed,
and there may be more visits to process. For example, if the last known archive for 2021-05-20 is calculated at 2021-05-20 20:00:00,

> but it is now 2021-05-21 00:30:00, there may be visits between 20:00:00 and 00:00:00 (midnight) to process.

Reviewer 2: Is that only true for `yesterday`? If archiving doesn't run for some days, I guess this will also be done for an older day, right?

Author: Currently only true for 'yesterday'. We assume core:archive runs at least once per day.

Reviewer 2: What happens if core:archive fails to run for a couple of days and is then started again?

Author: Then it should launch archiving for the week which will archive the days in between. This is not ideal but it is basically the same behavior as before, except then we did last52 days.

Author: Btw, this isn't meant to be a review of the archiving code but of this documentation which would hopefully allow you to understand and maintain the system.

Reviewer 2: Sure. Was just wondering how that is currently solved and if it's worth to mention

Author: I can add "doesn't run for a couple days" as a case here.

**invalidating custom ranges**

If custom ranges are specified in the `[General] archiving_custom_ranges` INI config setting and/or
the default period to load in Matomo is configured to be a range, they will be invalidated and
rearchived during `core:archive`.

**queued invalidations**

> Archives that are invalidated from system functions, for example, invalidating past data when a segment or entity changes, are

Reviewer: here we could also mention when tracking a request for an older date? Then we remember that archive to be invalidated.

Author: :+1:

  archives containing today are always eventually reprocessed. By contrast, for data that is tracked in the past, we
  explicitly invalidate the period. We don't do this for `today`, because invalidating data for every visit encountered
  today would make the tracker far less performant. Tracking data in the past, however, does not happen with nearly as
  high frequency or volume.

- if the period is a range, then we always recompute it when it is no longer valid. This is because ranges are just two
  arbitrary dates. We cannot preprocess every single permutation during core:archive, so instead we launch the archiving process
  in browser requests when the ttl expires. Since range archives just aggregate day periods together, and we don't launch
  archiving for those days, this is still performant.

The TTLs are determined by user and config settings. The day period can be specific in the UI in the System Settings page
and through INI config. The other periods can have custom TTLs in INI config as well, but not in the UI. See above for
the specific settings.

If no custom TTL is specified for a non-day period, it defaults to the day TTL.

**amount of concurrency to expect**

> In general, we can expect that individual core:archive processes will not process multiple sites at the same time. It is

Reviewer: I believe technically we can expect multiple sites to be launched at the same time? Even the same site.

Eg if there is only one site, then

* We launch archiver, queue will be eg `$idsites=[1]` and after starting command the queue will be empty
* Next archiver will launch the queue is empty and will be reset to `$idsites=[1]` and the same site will be archived.

Or do we maybe prevent this?

Same could happen from time to time when there are multiple sites... especially if one or few sites take longer to archive while others are faster then the queue will be empty and we reset the list of sites to archive etc.

Or maybe I'm not understanding what is meant by "will not process multiple sites at the same time"

Reviewer: actually I see now you're mentioning it re SharedSiteIds

Author: It mentions that it's possible w/ the current implementation and probably happens regularly if rarely, but it's not the main use case. And that we want things to work when that happens, but are not too concerned with the performance penalty from the edge case.

- has no visits in the log tables for the period
- and has no subperiod archive within the period (ie, if we're looking at a week and there are no day archives for that week, this would be true)

We know we'd end up with zero visits if we archive, so we can just skip it. The only exceptions are for sites that do
not use the tracker (so there will always be zero visits in the log table) or are specified via the `'Archiving.getIdSitesToArchiveWhenNoVisits'`
event. RollUp websites are examples of sites for which this optimization does not apply.

This optimization is used both before launching the archive aggregation logic and in the `core:archive` command
before we launch individual archive requests. This saves a bit more time since we also don't have to launch an
archiving command.

_If an archive's TTL is still valid_

This isn't truly an optimization, but if an archive's TTL is still valid, and it is for an archive where we check
the TTL (periods that include today and range periods), then we won't launch archiving. See below for more information
on archive TTLs.

### Special URL Query Parameter Handling

> * `trigger`: if set to `archivephp`, let's the core archiving process know it was launched by the `core:archive` command.

Reviewer 3: ```suggestion
* `trigger`: if set to `archivephp`, lets the core archiving process know it was launched by the `core:archive` command.
```

  to initiate archiving for all plugins.
  
  Note: for some archives (where period=range or if using a custom segment), we will only initiate archiving for the requested plugin(s).
- **Piwik\PluginsArchiver**: This class will loop through every plugin (or requested plugin), create the plugin's `Archiver` instance
  and invoke it. Based on whether a day or non-day period is being archived, either the `aggregateDayReports()` or `aggregateMultipleReports()`
  method will be invoked.
- **Piwik\Plugin\Archiver**: This is the base class for all plugin archivers. Plugins define their archiving logic within one of these
  classes, which will then be invoked by PluginsArchiver.
- **Piwik\ArchiveProcessor**: This is a utility class that contains methods to easily query and aggregate metrics and reports for multiple periods.
  It is mostly used by `aggregateMultipleReports()` implementations to aggregate and insert archive data with a single method call, and by
  both `aggregateMultipleReports()` and `aggregateDayReports()` to insert archive data.
- **Piwik\DataAccess\LogAggregator**: This is a utility class that contains some methods to aggregate log data. It is mostly used by

>   `aggregateDayReports()` implementations to aggregate log data and turn the result into `DataTable`s to insert. Many plugin may find

Reviewer 3: ```suggestion
  `aggregateDayReports()` implementations to aggregate log data and turn the result into `DataTable`s to insert. Many plugins may find
```

- **Piwik\Archive\Parameters**: This is a utility class that contains all the parameters of an archive data queries, such as the
  site IDs to get data for, the periods to get data for and the segment to get data for.
- **Piwik\ArchiveProcessor\Parameters**: This is a utility class that, like `Piwik\ArchiveParameters`, contains the parameters for
  a single archiving run. It only stores one site ID and period, however, since we can only launch the archiving process for a single site/period
  at a time. It also contains some other information like the request plugin/report if any or whether we want to only archive data for
  a single report.

The entire code path for Core Archiving is as follows:

- An API class constructs an `Archive` instance and queries for archive data.
- If archiving is not allowed to launch, `Archive` will look through the archive tables directly for data and return whatever it finds.
- If archiving is allowed, `Archive` will initiate archiving for every site/period combination in the query. It does this by calling the
  `CoreAdminHome.archiveReports` API method.
- `CoreAdminHome.archiveReports` creates and invokes the `Loader` class.

> - `Loader` looks for a usable archive. If found, it returns that archive ID. Otherwise it launches the archiving process, first aggregating

Reviewer 3: ```suggestion
- `Loader` looks for a usable archive. If found, it returns that archive ID. Otherwise, it launches the archiving process, first aggregating
```

  core metrics like nb_visits. Then if we can't skip the archive, it invokes `PluginsArchiver`.
- `PluginsArchiver` loops through every activated plugin and creates the plugin's `Archiver` instance if there is one. It invokes the appropriate
  method on the `Archiver` instances.
- `Archiver` instances define `aggregateDayReports()` and `aggregateMultipleReports()` methods. `aggregateDayReports()` methods will
  generally use a `LogAggregator` instance to aggregate log data and then insert it using an `ArchiveProcessor` instance. `aggregateMultipleReports()`
  methods will use an aggregation method in `ArchiveProcessor`.
  - `ArchiveProcessor` aggregation methods will query data using an internal `Archive` instance created using the current period's subperiods
    (so if the current period is a week, this `Archive` will query for days within that week).

>   - `ArchiveProcessor` will aggregate that data together ans insert it into the archive tables.

Reviewer 3: ```suggestion
  - `ArchiveProcessor` will aggregate that data together and insert it into the archive tables.
```

for each valid entry.

### PHP Classes

* **Piwik\Plugins\CoreConsole\Commands\CoreArchiver**: The `Command` class for the `core:archive` command. This will create and use a `CronArchive`
  instance. This is the main code path from which `CronArchive` is invoked, the only other being the `CoreAdminHome.runCronArchiving` API method.
* **Piwik\CronArchive**: Encapsulates the entire cron archiving process. This class will loop over every requested site (or work on an existing
  shared queue of site IDs), invalidate data tha needs to be invalidated for it, and process archives for every invalidation for the site.
* **Piwik\CronArchive\QueueConsumer**: This class is used by CronArchive to get the next invalidated archives to process. It queries the
  archive_invalidations table and returns a batch of archive invalidations to process. CronArchive will then launch API requests to
  `CoreAdminHome.archiveReports` for those invalidations.
* **Piwik\CliMulti**: This is a utility class used by CronArchive to initiate multiple API requests in parallel. It will try to execute the requests

>   from the CLI using the `climulti:request` command if it can. Otherwise it falls back to initiating curl requests.

Reviewer 3: ```suggestion
  from the CLI using the `climulti:request` command if it can. Otherwise, it falls back to initiating curl requests.
```

  we use what's there. `SharedSiteIds` will pull one site ID at a time from the option and give it to `CronArchive` to process. This allows
  multiple `core:archive` instances to work on the same list.
* **Piwik\CronArchive\SegmentArchiving**: This is a utility class that contains some logic for getting the stored segments to invoke archiving for
  and how long in the past we need to re-archive data for when a segment is created or updated.
* **Piwik\Archive\ArchiveInvalidator**: This is a core service class that is not strictly part of the core:archive system, but is very important
  to it. It contains the logic to invalidate archive data, and `CronArchive` uses it (indirectly) to invalidate archives before it starts processing
  for a single site.

The entire code path for core:archive execution is:

- The `CoreArchiver` command class is invoked. It processes user input and constructs a `CronArchive` instance. Options are set on the instance and
  it is invoked.
- `CronArchive` uses `FixedSiteIds` or `SharedSiteIds` to start pulling sites to archive. For every site in that list:

>   - `CronArchive` invalidates any data that needs to be invalidate for the site. This can be from visits tracked in the past, entities like segments that were

Reviewer 3: ```suggestion
  - `CronArchive` invalidates any data that needs to be invalidated for the site. This can be from visits tracked in the past, entities like segments that were
```

    created or updated or special periods specified in config or settings. The invalidations are inserted into the `archive_invalidations` table.
  - `CronArchive` then uses `QueueConsumer` to pull invalidations to process from the `archive_invalidations` table which pulls them in the correct execution
    order.
  - `CronArchive` takes the batch of invalidations from `QueueConsumer` and uses `CliMulti` to launch multiple archiving processes via API method.

>   - `CronArchive` looks at the results of those API calls and determines whether the request was successful. If so, the invalidation is removed. Otherwise

Reviewer 3: ```suggestion
  - `CronArchive` looks at the results of those API calls and determines whether the request was successful. If so, the invalidation is removed. Otherwise,
```

    it is unset and we ignore it until the next time we process this site.
  - This continues until `QueueConsumer` reports no invalidations to process.
- We pull a new site and process it. This continues until there are no sites to process.

### Range Archiving vs All Others

By default when using core:archive, day, week, month & year periods are not processed in the browser. Range periods
on the other hand are still processed from web requests. This is because we cannot pre-archive every possible permutation
of dates that users can specify.

> Instead when a range period is no longer considered up-to-date (it's TTL expires), we launch archiving for the range.

Reviewer 3: ```suggestion
Instead when a range period is no longer considered up-to-date (its TTL expires), we launch archiving for the range.
```
